server/game/effects.js

```diff
@@ -1064,6 +1064,17 @@ const Effects = {
             }
         };
     },
+    cannotBeDiscardedAtRandomByGameAction: function () {
+        return {
+            targetType: 'player',
+            apply: function (player) {
+                player.cannotBeDiscardedAtRandom = true;
+            },
+            unapply: function (player) {
+                player.cannotBeDiscardedAtRandom = false;
+            }
+        };
+    },
```

Comment:
I wouldn't be adding a new effect here (and checking `cannotBeDiscardedAtRandom` within `DiscardAtRandom.canChangeGameState`), as this is ultimately somethign that should be properly handled by `GameAction.allow`, where it checks the name of the game action (in this case, "discardAtRandom") can legally be performed on that player.

I would recommend opening a debug session, starting from `this.game.resolveGameAction` on `HeadsOnSpikes.js` and getting to the point of checking that `allow` function, and seeing why it's not recognising that the action cannot be performed on the Laughing Storm player.